packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -609,11 +609,60 @@ class WidgetTester extends WidgetController implements HitTestDispatcher, Ticker
   /// are required to wait for the returned future to complete before calling
   /// this method again. Attempts to do otherwise will result in a
   /// [TestFailure] error being thrown.
+  ///
+  /// If your test code depends on the result of a task that did not complete,
+  /// the most common symptom would be a hanging test as FakeAsync environment
+  /// is unable to resolve a future that was created in `runAsync`. If you
+  /// observe such behavior or flakiness, you have a number of options:
+  ///
+  /// - Consider restructuring your code so you do not need `runAsync`. This is
+  ///   the optimal solution as widget tests are designed to run in Fake async
+  ///   environment.
+  ///
+  /// - Expose a [Future] in your application code that signals the readiness of
+  ///   your widget tree, then await that future inside `callback`.
+  ///
+  /// - If either option is not feasible, await [resolveFutures] inside
+  ///   `callback`.
   Future<T> runAsync<T>(
     Future<T> callback(), {
     Duration additionalTime = const Duration(milliseconds: 1000),
   }) => binding.runAsync<T>(callback, additionalTime: additionalTime);
 
+  /// Waits for a stable state within a single `runAsync` execution.
+  ///
+  /// Note that using this method is *not* recommended. This method should only
```

Comment:
Adding API that we do not recommend to be used seems ... wrong.

Reply:
I would also love to add "not recommended" to `runAsync` FWIW :-)

Reply:
what's wrong with runAsync?

Reply:
- Zones.. testWidgets() has two zones, which itself is tough to grasp. runAsync within testWidgets ups that to three. Most developers don't understand zones and the side effects are bizarre and difficult to debug.
- Makes it too easy to run into synchronization problems which is another thing that's difficult to debug. 

Sorry for being generic. I can probably compile concrete examples for these. These are some of the difficulties I myself have run into and seen other developers struggle with.